website/content/en/docs/upgrading/v1-migration.md

```diff
@@ -176,233 +289,297 @@ kubectl patch customresourcedefinitions ec2nodeclasses.karpenter.k8s.aws -p "{\"
-   However, compatibility is not guaranteed long-term and changes to either AL2 or Ubuntu's UserData format may introduce incompatibilities.
-   If this occurs, the Custom AMIFamily should be used for Ubuntu and UserData will need to be entirely maintained by the user.
+   {{% /alert %}}
 
-   If you are upgrading to `v1.0` and already have v1beta1 Ubuntu EC2NodeClasses, all you need to do is specify `amiSelectorTerms` and Karpenter will translate your NodeClasses to the v1 equivalent (as shown below).
-   Failure to specify `amiSelectorTerms` will result in the EC2NodeClass and all referencing NodePools to show as NotReady, causing Karpenter to ignore these NodePools and EC2NodeClasses for Provisioning and Drift.
+3. Attach the `v1beta1` policy from your target version to your existing NodeRole.
+   If you didn't remove the `v1beta1` policy after upgrading to v1, you may skip this step.
 
-   ```yaml
-   # Original v1beta1 EC2NodeClass
-   version: karpenter.k8s.aws/v1beta1
-   kind: EC2NodeClass
-   spec:
-     amiFamily: Ubuntu
-     amiSelectorTerms:
-     - id: ami-foo
-   ---
-   # Conversion Webhook Output
-   version: karpenter.k8s.aws/v1
-   kind: EC2NodeClass
-   metadata:
-     annotations:
-       compatibility.karpenter.k8s.aws/v1beta1-ubuntu: amiFamily,blockDeviceMappings
-   spec:
-     amiFamily: AL2
-     amiSelectorTerms:
-     - id: ami-foo
-     blockDeviceMappings:
-     - deviceName: '/dev/sda1'
-       rootVolume: true
-       ebs:
-         encrypted: true
-         volumeType: gp3
-         volumeSize: 20Gi
+   ```bash
+   TEMPOUT=$(mktemp)
+   VERSION_TAG=$([[ ${KARPENTER_VERSION} == v* ]] && echo "${KARPENTER_VERSION}" || echo "v${KARPENTER_VERSION}")
+   curl -fsSL https://raw.githubusercontent.com/aws/karpenter-provider-aws/${VERSION_TAG}/website/content/en/preview/getting-started/getting-started-with-karpenter/cloudformation.yaml > ${TEMPOUT}
+   sed -e 's/!Sub//g' -i "" "${TEMPOUT}"
+   sed -e 's/${AWS::Partition}/${AWS_PARTITION}/g' -i "" "${TEMPOUT}"
+   sed -e 's/${AWS::Region}/${AWS_REGION}/g' -i "" "${TEMPOUT}"
+   sed -e 's/${AWS::AccountId}/${AWS_ACCOUNT_ID}/g' -i "" "${TEMPOUT}"
+   sed -e 's/${ClusterName}/${CLUSTER_NAME}/g' -i "" "${TEMPOUT}"
+   sed -e 's/${KarpenterInterruptionQueue.Arn}/arn:${AWS_PARTITION}:sqs:${AWS_REGION}:${AWS_ACCOUNT_ID}:${CLUSTER_NAME}/g' -i "" "${TEMPOUT}"
+   sed -e 's/${KarpenterNodeRole.Arn}/arn:${AWS_PARTITION}:iam::${AWS_ACCOUNT_ID}:role\/KarpenterNodeRole-${CLUSTER_NAME}/g' -i "" "${TEMPOUT}"
+
+   POLICY_DOCUMENT=$(yq '.Resources.KarpenterControllerPolicy.Properties.PolicyDocument' ${TEMPOUT} | envsubst)
+   POLICY_NAME="KarpenterControllerPolicy-${CLUSTER_NAME}-${VERSION_TAG}"
+   ROLE_NAME="${CLUSTER_NAME}-karpenter"
+   POLICY_ARN="$(aws iam create-policy --policy-name "${POLICY_NAME}" --policy-document "${POLICY_DOCUMENT}" | jq -r .Policy.Arn)"
+   aws iam attach-role-policy --role-name "${ROLE_NAME}" --policy-arn "${POLICY_ARN}"
```

Comment:
Is this all in the cloudformation template already?

Reply:
Kind of, this attaches the policy from the `v1` template to the NodeRole. We'll upgrade the CFN stack in the final step and remove this temporary policy. This just helps minimize downtime by ensuring Karpenter has the proper permissions for both the source and target version while migrating.